Cleanup and refactor Manage Permissions

rdig opened the pull request:
## Description

Cleans up imports and refactors the `managePermissions` saga, adds and refactors Schema Models, wires up the Actions list and page.

I've left comments where relevant, especially in the Schema file, as the changes to the models were less than obvious.

### Screenshots

![Screenshot from 2023-04-24 07-52-24](https://user-images.githubusercontent.com/1193222/233906015-589c540b-dd03-4ef1-b7c3-ff20d0f3d84e.png)
![Screenshot from 2023-04-24 07-52-30](https://user-images.githubusercontent.com/1193222/233906020-bd09f1f2-995c-411d-9b01-94503b76a3c0.png)

### Testing

- Pull everything
- Start the dev enviroment
- create your user 
- create your colony _(I've noticed that the first ever colony created on the chain acts weird, and the ingestor doesn't always pick up it's events, so I'd recomment actually creating a second one, at which point it will work flawlessly)_
- Press `Test Permissions #1`
- Press `Test Permissions #2`
- Refresh the page 
- You should now see / click various actions entries for permissions _(Note that the entry that gives permissions to the extension, actually says that it gives permissions to the colony. This is visually "wrong", the data is all there, but the `FriendlyName` component doesn't know how to display it -- yet)_

Resolves #234
Resolves #235 
Resolves #236
Resolves #237

jakubcolony commented on amplify/backend/api/colonycdapp/schema.graphql:
```diff
@@ -322,6 +322,10 @@ type Colony @model {
   version: Int!
   actions: [ColonyAction] @hasMany
   metadata: ColonyMetadata @hasOne(fields: ["id"])
+  # @TODO This should not be fetched upfront
+  # It should be retrieved on demand at the earliest occasion it's needed
+  # ie: when opening a UAC modal
+  roles: [ColonyRole] @hasMany
```
Would this comment be more appropriate in the CDapp where we actually fetch it?

rdig replied:
I actually don't know why I left a todo here? But the comment should be also here, although I agree I need to move it to the Dapp as well

jakubcolony commented on src/graphql/fragments/actions.graphql:
```diff
@@ -8,10 +8,28 @@ fragment ColonyAction on ColonyAction {
   initiatorUser {
     ...User
   }
+  initiatorColony {
+    ...Colony
+  }
+  initiatorExtension {
+    ...Extension
+  }
+  initiatorToken {
+    ...Token
+  }
```
Do we really need to query all the fields of those, or should a more restricted fragments be created? Getting the `Colony` fragment after the `balances` were included already took a performance hit and I feel like this would slow down the action query quite a bit if any of these are not null

rdig replied:
Good point. I think we need Fragments for these four (User, Colony, Extension, Token) in the "friendly name" format. Ie: just `name`, `displayName`, `address`

jakubcolony replied:
This is still outstanding I believe and potentially more urgent than my other comments

jakubcolony commented on amplify/backend/api/colonycdapp/schema.graphql:
```diff
@@ -501,4 +513,99 @@ type ColonyAction @model {
+      sortKeyFields: ["createdAt"]
+    )
+
+  domain: Domain! @hasOne
+  colony: Colony! @hasOne
```
Nitpick, but I like to manually define the foreign key fields for these (e.g. add `colonyId` of type `ID` and pass it to the `@hasOne` decorator), as otherwise the generated names just end up a bit confusing:
<img width="286" alt="image" src="https://user-images.githubusercontent.com/112586815/234130036-61468c2d-88bd-4ce6-8a3c-7492af581e6d.png">

rdig replied:
Good idea, didn't think about it.

jakubcolony commented on amplify/backend/api/colonycdapp/schema.graphql:
```diff
@@ -501,4 +513,99 @@ type ColonyAction @model {
+  role_6: Boolean # administration
+}
+
+# it's a model, not a type since this is designed to work outside of the main
+# colony model, or colony role model, as, by design it should be only used
+# in very specific cases
```
I don't think I know enough but why does it need to work outside of the colony role model?

rdig replied:
Outside, as it not directly linked (via @hasOne @hasMany, etc) as this is envisioned to just be fetched on an instance by instance basis. 

This will become relevant for Recovery Mode.

jakubcolony reviewed (commented):
Great job navigating around all the obstacles! 

After starting the dev env, I'm seeing a bunch of Amplify mutation errors:
<img width="598" alt="image" src="https://user-images.githubusercontent.com/112586815/234127823-c01282f7-4aab-4960-b3bd-994641cba3d8.png">

Then when I create a colony:
<img width="537" alt="image" src="https://user-images.githubusercontent.com/112586815/234131527-ae15c26b-424d-44c1-8030-5c23e4835255.png">

There are some actions in the list, so I don't know if the above breaks anything. The actions look duplicated however (I haven't clicked on either of the buttons, it's a fresh colony):
<img width="762" alt="image" src="https://user-images.githubusercontent.com/112586815/234131773-e9dec645-f20c-4e95-873c-cbafd28dcf2e.png">

Looking at the Dapp, there seems to only have been one action related to permissions on colony creation.

Both buttons work and the actions they trigger show correct values:
<img width="1036" alt="image" src="https://user-images.githubusercontent.com/112586815/234132153-c4dcd1ff-15a9-48d9-b8d5-f3181ec9bfff.png">

jakubcolony commented on amplify/backend/api/colonycdapp/schema.graphql:
```diff
@@ -501,4 +513,99 @@ type ColonyAction @model {
   newColonyVersion: Int
+  roles: ColonyActionRoles
+
+  # Required since some actions might have multiple event entries in the list
+  # which need to be iterated over and exposed individually
+  #
+  # This should be a list/array, but this is yet another thing Amplify is bad at
+  # The official response from them is to use DynamoDB directly:
+  # https://stackoverflow.com/questions/50992106/append-item-to-list-using-aws-appsync-to-dynamodb
+  # Which is just freaking great!
+  #
+  # As such, I've decided to simplify this, an just pass a stringified JSON around
+  # since this is for display purpouses only.
+  #
+  # If anyone has a better idea on how to handle this, I'll gladly change it!
+  #
+  # When / if you decide to use it, just be mindful of what format other people
+  # are using, and follow suit.
+  #
+  # Roles Types:
+  # [{ id: String, type: String, role: Number, setTo: boolean }]
```
Circling back to it, I realised I actually didn't look into individual events as an actions concept at all. Should we come up with a solution that caters for all actions?
Otherwise maybe this should be renamed to be more roles-specific?

rdig replied:
Yes we should, I'm not particularly happy with my implementation here, I just did it since I wanted something fast to get it done already :)

rdig replied:
That's why I asked @willm30 about his implementation on Motions, as to try to align either this to his, or come up with a middle ground.

willm30 commented on src/components/common/ColonyActions/ActionDetailsPage/ActionDetailsPageFeed/ActionDetailsPageFeed.tsx:
```diff
@@ -17,14 +17,17 @@ const ActionDetailsPageFeed = ({
 ActionsPageFeedProps) => {
-  const events = ACTIONS_EVENTS[actionData.type];
+  const events =
+    JSON.parse(actionData.individualEvents as string) ||
```
I would leave a comment here explaining why this is safe (i.e. how you know that actionData.individualEvents won't be undefined and cause the JSON.parse to error)

rdig replied:
Sure, but I'm not going to bother with this since we're going to change this logic anyway.

willm30 commented on amplify/backend/api/colonycdapp/schema.graphql:
```diff
@@ -501,4 +513,99 @@ type ColonyAction @model {
+  #
+  # Roles Types:
+  # [{ id: String, type: String, role: Number, setTo: boolean }]
+  individualEvents: String
```
I think I'm suggesting the same approach as Jakub above, but with motion messages all we do is fetch the current array, append to it, and then overwrite it. 

I would prefer that approach to this (type safety, consistency)

rdig replied:
As I've mentioned above, I'm not particularly happy with this, and will gladly change it to something better.

Shall we set up a call ? @willm30 @jakubcolony ?

willm30 commented on amplify/backend/api/colonycdapp/schema.graphql:
```diff
@@ -501,4 +513,99 @@ type ColonyAction @model {
+  role_1: Boolean # root
+  role_2: Boolean # arbitration
+  role_3: Boolean # architecture
+  role_5: Boolean # funding
```
Perhaps leave a comment here too as to why 4 is missing

willm30 commented on amplify/backend/api/colonycdapp/schema.graphql:
```diff
@@ -501,4 +513,99 @@ type ColonyAction @model {
+
+type ColonyRole @model {
+  id: ID! @index # colonyAddress_domainNativeId_userAddress_roles
+  domain: Domain! @hasOne
```
Nitpick: it's not clear which of these params are variable and constant in this comment. 

E.g. would be clearer:
``` graphql
# <colonyAddress>_<domainNativeId>_<contractAddress>_roles
```

willm30 replied:
Also, is there a reason for marking this as:

`id: ID! @index` ?

It's the primary key by default so I'm not sure what we gain from adding the `@index`

willm30 commented on amplify/backend/api/colonycdapp/schema.graphql:
```diff
@@ -501,4 +513,99 @@ type ColonyAction @model {
+  # @NOTE Always set to `type` to "SortedHistoricRole" to enable Amplify sorting, it DOES NOT WORK with `id`
+  #
+  # Apparently, it can only filter non-primary fields, meaning that if we sort by `blockNumber` then that
+  # field is considered primary, so it won't allow us to filter by it.
```
What is the intended access pattern here? 

If you wanted to get permissions at an exact block number, you could also have block number as an index.

Or if you wanted to get all permissions for a given target address, index the address and then you can have the block number as the sort key. 

We _probably_ don't need to add an abritrary field just to enable the desired access pattern (although hard to comment for sure since I'm not sure what it is).

willm30 replied:
If you indexed the targetAddress, and added the block number as a sortKeyField, then you'd be able to get all roles at the targetAddress and the block number without having to do any filtering. 

Then you'd only have to filter by colonyAddress to enable this pattern:

> get the permissions for user X at block Y within colony Z

This would also allow to you to get Roles for a target address from a  range block numbers since the sort key enables a variety of comparisons (gt, lt, gt & lt etc)

willm30 commented on amplify/backend/api/colonycdapp/schema.graphql:
```diff
@@ -501,4 +513,99 @@ type ColonyAction @model {
   toDomain: Domain @hasOne(fields: ["toDomainId"])
   fundamentalChainId: Int
   newColonyVersion: Int
+  roles: ColonyActionRoles
```
Nitpick: couldn't this point to the ColonyHistoricRoles entry created at the same time? They contain the same info, no?

rdig replied:
So these are different from the Historic Roles, this just has the exact roles that changed vs. historic role which have the exact snapshot at the time of them being set.

Eg: 
Action Role:
```
role_0: true
role_1: false
```
vs
```
role_0: true
role_1: false
role_2: null
role_3: null
role_5: null
role_6: null
```

willm30 commented on src/components/common/ColonyActions/ActionDetailsPage/ActionDetailsPageFeed/ActionDetailsPageFeed.tsx:
```diff
@@ -17,14 +17,17 @@ const ActionDetailsPageFeed = ({
-  const events = ACTIONS_EVENTS[actionData.type];
+  const events =
+    JSON.parse(actionData.individualEvents as string) ||
+    ACTIONS_EVENTS[actionData.type];
```
I think this should be refactored so all action events are of the same type (instead of having some as just strings and others like this as objects). 

At the very least we want:

``` graphql

type ActionEvent {
 name: String!
}
```
And then we can add all the rest as optional.

That would mean you could pass the event down into ActionsPageEvent directly, so you wouldn't then have to "find" it again inside `getColonyRoleSetTitleValues`

willm30 replied:
Maybe not in this issue but it'd be nice to make this a bit more coherent

willm30 reviewed (commented):
I am also receiving many amplify/ingestor errors of the following nature after building the branch for the first time (before doing anything): 

![errs](https://user-images.githubusercontent.com/64402732/234259840-c4e88821-6265-4b09-b69e-4a2ed289daf2.png)

rdig commented:
This is ready for re-review/merge.

However do note, that in the interest of saving as much time as possible for the impeding launch, I've chosen to not implement _(currently)_ some of the more time consuming improvements that were pointed out.

They are still relevant, so I've extracted them in their own separate improvement issue #579 which should/will be tackled once we launch, in the subsequent fixes deployments

willm30 commented:
I'm noticing a few issues that appear to be from the rebase:
1. `getRolesForUserAndDomain` - import is wrong
2. `getExtendedActionType` is missing
3. roles is commented out in `getDetailItems`
4. recipient is not defined in the ColonyActionFragment -> needs to be recipientAddress and references to recipient.walletAddress need to be replaced

I also needed to change my local appsync url in the block ingestor `.env` to  `http://localhost:20002/graphql` (previously it was just localhost:20002) to stop the block ingestor from erroring.
 
If I fix these manually, things appear to work as intended. I haven't noticed that there's anything I can't do as the creator of colony (i.e. all permissions enabled).  

I have noticed a few things I shouldn't be able to do. If I switch user to one without permissions, I can: 

- click "install" on the voting rep extn (which then fails). 
- submit the manage permissions form if I assign this user just the root permission (where you need permission 3 also). 

Will we leave these to be fixed in QA? 

The ActionDetailsPage looks as expected, aside from the message: "RecoveryRoleSet emitted by". Looks like it's missing an address.

![permissions-screen](https://github.com/JoinColony/colonyCDapp/assets/64402732/94c82da6-f3d6-48b0-8caf-f75075060b58)

Happy to approve and fix the rest in qa once the rebase issues are fixed.

Julianahlk reviewed (commented):
I tried making all the fixes @willm30 mentioned but I am still getting an error every time I try to create a colony.

<img width="1456" alt="Screen Shot 2023-05-24 at 4 21 19 PM" src="https://github.com/JoinColony/colonyCDapp/assets/71563622/e421a1df-a9bd-4b93-a309-043ccd17f8f5">

rdig commented:
@willm30 @Julianahlk Fix everything you guys pointed out, and then some, as most of the dialogs were not wired with permissions at all

Some notes on testing:
- Due to QA changes you'll need to update your `.env` file permanently for the GraphQL link to include the `/graphql` part of the url _(as @willm30 correctly deduced)_ -- this is going to make it's way into `master` eventually and it's going to be a requirement from now on
- permissions preventing you to unlock token is correct as that's not wired in yet _(not in this PR at least -- I do remember it being available on the QA branch though)_

@willm30 

> submit the manage permissions form if I assign this user just the root permission (where you need permission 3 also).

This is actually correct the way it's currently implemented. You need Root in the Top Domain (Root Domain), or Architecture in the subdomains

willm30 commented:
Unlock tokens should be available, dialog + saga were merged to master in these prs:
https://github.com/JoinColony/colonyCDapp/pull/217
https://github.com/JoinColony/colonyCDapp/pull/320

I'm also not able to mint tokens in a colony, despite having root permissions.

![mint-tokens](https://github.com/JoinColony/colonyCDapp/assets/64402732/5e94cbca-22a7-49a3-8204-a93cb8aa50f8)

According to this tooltip, if I have a permission in root, I should also have it in a subteam. I have created subteam `123` in Root, but the Payment dialog informs me I don't have the funding and administration permissions, despite having them in the root domain.

![subteams](https://github.com/JoinColony/colonyCDapp/assets/64402732/0f5f13f6-98ce-4e43-8a8b-823d8ad602cc)

> This is actually correct the way it's currently implemented. You need Root in the Top Domain (Root Domain), or Architecture in the subdomains

I have a user with only root permission. When I go to assign that user more permissions in root, I get the following chain error:

![managepermissions](https://github.com/JoinColony/colonyCDapp/assets/64402732/b94abdfa-4fee-49b3-ada3-a316f8144636)

Julianahlk reviewed (commented):
Overall, seeing the same issues as Will. One thing to add is in the Manage Permissions Dialog, when I view the permissions for a subteam with a user that has all the permissions in Root, the tooltip says I do not have the permission even though the permission is checked.

<img width="496" alt="Screenshot 2023-05-31 at 1 57 27 PM" src="https://github.com/JoinColony/colonyCDapp/assets/71563622/48c9126d-594a-4235-a684-ace312f0dad2">

rdig commented:
@willm30 @Julianahlk Did another round of fixes based on your findings, so please give it another look

@willm30 

> Unlock tokens should be available, dialog + saga were merged to master in these prs:

Those PRs do not wire in the `colony.status.nativeToken` logic that the modals are basing their logic check on. 

I do remember however that there was some work towards that by @ArmandoGraterol so maybe he can shed some light here.

> I'm also not able to mint tokens in a colony, despite having root permissions.

Same deal as above, it checks for `colony.status.nativeToken.mintable` which is not available

> According to this tooltip, if I have a permission in root, I should also have it in a subteam. I have created subteam 123 in Root, but the Payment dialog informs me I don't have the funding and administration permissions, despite having them in the root domain.

Good spot, as well as correct. Fixed.

> I have a user with only root permission. When I go to assign that user more permissions in root, I get the following chain error:

> I would expect the form to be disabled with a message in the ui about missing permission 3.

Correct as well, seems I've misunderstood the first time you pointed this out. This was just a case of pointing the logic to the correct permissions. Fixed

---

@Julianahlk 

> When I view the permissions for a subteam with a user that has all the permissions in Root, the tooltip says I do not have the permission to set the role.

Fixed alongside the subteam logic I fixed in Will's comment above _(was a issue with `getUserRolesForDomain`'s fallbakc logic)_

jakubcolony commented on src/components/common/ColonyActions/helpers/mapItemToMessageFormat.tsx:
```diff
@@ -87,26 +94,28 @@ export const mapColonyActionToExpectedFormat = (
     ),
-    // ...getColonyRoleSetTitleValues(role?.setTo),
+    ...getColonyRoleSetTitleValues(
+      actionData?.individualEvents as string,
```
Optional chaining looks like it's not needed here, and could you provide a comment why the casting to string is safe here (otherwise this could be null or undefined as far as the type is concerned)?

jakubcolony commented on src/utils/colonyActions.ts:
```diff
@@ -487,6 +488,28 @@ export const getDetailsForAction = (
+    { id: 6, setTo: roles.role_6 },
+  ];
+  return extractedRoles.filter(
+    ({ setTo }) => setTo !== null,
```
Just making sure this is right, you're not excluding the case in which `setTo` is undefined (and looking at the type it could be)

jakubcolony reviewed (approved):
Left a few comments but otherwise LGTM, great effort!
